File: scripts/scrape_eporner.py
#!/usr/bin/env python3
"""
Eporner Web Scraper Tool
Scrapes eporner.com to extract HTML/CSS structure for styling reference
"""

import requests
from bs4 import BeautifulSoup
import json
import re
import os
from urllib.parse import urljoin

class EpornerScraper:
    def __init__(self):
        self.base_url = "https://www.eporner.com"
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Connection': 'keep-alive',
        }
        self.session = requests.Session()
        self.session.headers.update(self.headers)
        
    def get_page(self, url):
        """Fetch a page from eporner.com"""
        try:
            response = self.session.get(url, timeout=30)
            response.raise_for_status()
            return BeautifulSoup(response.text, 'html.parser')
        except requests.RequestException as e:
            print(f"Error fetching {url}: {e}")
            return None
    
    def get_page_title(self, soup):
        """Get page title"""
        if soup.title:
            return soup.title.string
        return None
    
    def extract_css_files(self, soup):
        """Extract all CSS file URLs"""
        css_files = []
        for link in soup.find_all('link', rel='stylesheet'):
            href = link.get('href')
            if href:
                css_files.append(href)
        return css_files
    
    def extract_main_styles(self, soup):
        """Extract inline styles from style tags"""
        styles = []
        for style in soup.find_all('style'):
            if style.string:
                styles.append(style.string)
        return styles
    
    def scrape_homepage(self):
        """Scrape homepage structure"""
        print("=" * 60)
        print("SCRAPING EPORNER HOMEPAGE")
        print("=" * 60)
        
        soup = self.get_page(self.base_url)
        if not soup:
            return None
            
        print(f"\n📌 Page Title: {self.get_page_title(soup)}")
        
        # Extract CSS files
        print(f"\n📄 CSS Files:")
        css_files = self.extract_css_files(soup)
        for i, css in enumerate(css_files[:10], 1):
            print(f"  {i}. {css}")
        
        # Extract main navigation
        print(f"\n🔗 Main Navigation:")
        nav = soup.find('nav')
        if nav:
            for link in nav.find_all('a', href=True)[:10]:
                print(f"  - {link.get('text', '').strip()}: {link['href']}")
        
        # Extract hero section
        print(f"\n🎯 Hero Section:")
        hero = soup.find('section', class_=re.compile(r'hero|banner|featured'))
        if hero:
            print(f"  Hero element found: {hero.name}")
            print(f"  Classes: {hero.get('class', [])}")
        
        # Extract video grid structure
        print(f"\n🎬 Video Grid:")
        video_containers = soup.find_all(['div', 'section'], class_=re.compile(r'video|grid|item|thumb'))
        print(f"  Found {len(video_containers)} video-related elements")
        
        # Extract categories
        print(f"\n📂 Categories:")
        category_links = soup.find_all('a', href=re.compile(r'/category|/cat'))
        for cat in category_links[:8]:
            print(f"  - {cat.get('text', '').strip()}: {cat.get('href')}")
        
        return {
            'title': self.get_page_title(soup),
            'css_files': css_files,
            'navigation': str(nav) if nav else None,
        }
    
    def scrape_video_page(self):
        """Scrape video watch page structure"""
        print("\n" + "=" * 60)
        print("SCRAPING EPORNER VIDEO PAGE")
        print("=" * 60)
        
        # First get a video URL from homepage
        soup = self.get_page(self.base_url)
        if not soup:
            return None
            
        # Find a video link
        video_link = soup.find('a', href=re.compile(r'/video/'))
        if not video_link:
            print("No video link found")
            return None
            
        video_url = urljoin(self.base_url, video_link['href'])
        print(f"Video URL: {video_url}")
        
        soup = self.get_page(video_url)
        if not soup:
            return None
            
        print(f"\n📌 Page Title: {self.get_page_title(soup)}")
        
        # Video player
        print(f"\n🎬 Video Player:")
        player = soup.find(['div', 'iframe', 'video'], class_=re.compile(r'player|video|embed'))
        if player:
            print(f"  Player found: {player.name}")
            print(f"  Classes: {player.get('class', [])}")
        
        # Video info
        print(f"\n📊 Video Info:")
        title = soup.find('h1')
        if title:
            print(f"  Title: {title.get_text(strip=True)}")
        
        # Views, rating, duration
        stats = soup.find_all(['span', 'div'], class_=re.compile(r'views|rating|duration|time'))
        for stat in stats[:6]:
            print(f"  {stat.get_text(strip=True)}")
        
        # Related videos
        print(f"\n📺 Related Videos:")
        related = soup.find_all('a', href=re.compile(r'/video/'))
        print(f"  Found {len(related)} video links")
        
        return {'video_url': video_url}
    
    def scrape_category_page(self):
        """Scrape category page structure"""
        print("\n" + "=" * 60)
        print("SCRAPING EPORNER CATEGORY PAGE")
        print("=" * 60)
        
        # Try a category URL
        category_url = f"{self.base_url}/categories"
        soup = self.get_page(category_url)
        if not soup:
            # Try alternative
            category_url = f"{self.base_url}/cat"
            soup = self.get_page(category_url)
            if not soup:
                return None
            
        print(f"📌 Page Title: {self.get_page_title(soup)}")
        
        # Category grid
        print(f"\n📂 Category Grid:")
        categories = soup.find_all('a', href=re.compile(r'/cat/|/category/'))
        print(f"  Found {len(categories)} category links")
        
        # Print first few categories
        for cat in categories[:6]:
            name = cat.get_text(strip=True)
            href = cat.get('href', '')
            if name and 'category' in href.lower():
                print(f"  - {name}: {href}")
        
        return {'category_count': len(categories)}
    
    def scrape_pornstars_page(self):
        """Scrape pornstars page structure"""
        print("\n" + "=" * 60)
        print("SCRAPING EPORNER PORNSTARS PAGE")
        print("=" * 60)
        
        pornstars_url = f"{self.base_url}/pornstars"
        soup = self.get_page(pornstars_url)
        if not soup:
            return None
            
        print(f"📌 Page Title: {self.get_page_title(soup)}")
        
        # Pornstar grid
        print(f"\n👤 Pornstar Grid:")
        pornstars = soup.find_all('a', href=re.compile(r'/pornstar/|/model/'))
        print(f"  Found {len(pornstars)} pornstar links")
        
        # Print first few
        for ps in pornstars[:6]:
            name = ps.get_text(strip=True)
            href = ps.get('href', '')
            if name and 'pornstar' in href.lower():
                print(f"  - {name}: {href}")
        
        return {'pornstar_count': len(pornstars)}
    
    def extract_css_variables(self, soup):
        """Extract CSS custom properties (variables)"""
        print("\n" + "=" * 60)
        print("EXTRACTING CSS VARIABLES")
        print("=" * 60)
        
        # Look for CSS in style tags
        styles = soup.find_all('style')
        css_vars = []
        
        for style in styles:
            if style.string:
                # Find CSS custom properties
                vars_found = re.findall(r'--[\w-]+:\s*[^;]+;', style.string)
                css_vars.extend(vars_found)
        
        # Also check inline styles
        for elem in soup.find_all(style=True):
            inline = elem.get('style', '')
            vars_found = re.findall(r'--[\w-]+:\s*[^;]+;', inline)
            css_vars.extend(vars_found)
        
        print(f"\n🎨 Found {len(css_vars)} CSS variables:")
        unique_vars = list(set(css_vars))[:20]
        for var in unique_vars:
            print(f"  {var}")
        
        return unique_vars
    
    def generate_css_reference(self):
        """Generate a reference file of eporner's CSS structure"""
        print("\n" + "=" * 60)
        print("GENERATING CSS REFERENCE")
        print("=" * 60)
        
        soup = self.get_page(self.base_url)
        if not soup:
            return
            
        # Get all CSS
        css_files = self.extract_css_files(soup)
        inline_styles = self.extract_inline_styles(soup)
        
        # Generate reference
        reference = {
            'base_url': self.base_url,
            'css_files': css_files,
            'main_colors': self.extract_main_colors(soup),
            'layout_structure': self.analyze_layout(soup),
        }
        
        # Save to file
        output_file = 'eporner_reference.json'
        with open(output_file, 'w') as f:
            json.dump(reference, f, indent=2)
        print(f"\n💾 Reference saved to: {output_file}")
        
        return reference
    
    def extract_inline_styles(self, soup):
        """Extract common inline style patterns"""
        patterns = {}
        
        for elem in soup.find_all(style=True):
            style = elem.get('style', '')
            # Extract common properties
            if 'background' in style:
                patterns['background'] = style
            if 'color:' in style:
                patterns['color'] = style
            if 'font' in style:
                patterns['font'] = style
        
        return patterns
    
    def extract_main_colors(self, soup):
        """Extract main color values from the page"""
        colors = set()
        
        # Check style tags
        for style in soup.find_all('style'):
            if style.string:
                # Find hex colors
                hex_colors = re.findall(r'#[0-9a-fA-F]{3,8}', style.string)
                colors.update(hex_colors[:10])
                
                # Find rgb colors
                rgb_colors = re.findall(r'rgb\(\d+,\s*\d+,\s*\d+\)', style.string)
                colors.update(rgb_colors[:5])
        
        # Check inline styles
        for elem in soup.find_all(style=True):
            style = elem.get('style', '')
            hex_colors = re.findall(r'#[0-9a-fA-F]{3,8}', style)
            colors.update(hex_colors[:5])
        
        return list(colors)[:20]
    
    def analyze_layout(self, soup):
        """Analyze the page layout structure"""
        layout = {
            'header': None,
            'main_content': None,
            'footer': None,
            'sidebar': None,
        }
        
        # Find header
        header = soup.find('header')
        if header:
            layout['header'] = {
                'tag': 'header',
                'classes': header.get('class', []),
                'id': header.get('id'),
            }
        
        # Find main
        main = soup.find('main')
        if main:
            layout['main_content'] = {
                'tag': 'main',
                'classes': main.get('class', []),
            }
        
        # Find footer
        footer = soup.find('footer')
        if footer:
            layout['footer'] = {
                'tag': 'footer',
                'classes': footer.get('class', []),
            }
        
        # Find sidebar
        sidebar = soup.find(['aside', 'div'], class_=re.compile(r'sidebar|aside'))
        if sidebar:
            layout['sidebar'] = {
                'tag': sidebar.name,
                'classes': sidebar.get('class', []),
            }
        
        return layout


def main():
    scraper = EpornerScraper()
    
    print("🚀 Starting Eporner Web Scraper")
    print("=" * 60)
    
    # Scrape different pages
    scraper.scrape_homepage()
    scraper.scrape_video_page()
    scraper.scrape_category_page()
    scraper.scrape_pornstars_page()
    
    # Generate reference
    reference = scraper.generate_css_reference()
    
    print("\n" + "=" * 60)
    print("✅ SCRAPING COMPLETE")
    print("=" * 60)
    print("""
Next steps:
1. Check eporner_reference.json for extracted data
2. Review the CSS structure
3. Apply the styling to your website
4. Run: python scripts/scrape_eporner.py
    """)


if __name__ == "__main__":
    main()

File: scripts/scrape_eporner.js
/**
 * Eporner Web Scraper Tool (Node.js)
 * Scrapes eporner.com to extract HTML/CSS structure for styling reference
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');

// Simple HTTP request helper
function fetch(url) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    const client = parsedUrl.protocol === 'https:' ? https : http;
    
    const options = {
      hostname: parsedUrl.hostname,
      path: parsedUrl.pathname + parsedUrl.search,
      method: 'GET',
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
      }
    };

    const req = client.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => resolve(data));
    });

    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });
    req.end();
  });
}

// Simple HTML parser (basic regex-based)
function parseHTML(html) {
  return {
    html,
    // Find title
    getTitle() {
      const match = html.match(/<title[^>]*>([^<]+)<\/title>/i);
      return match ? match[1].trim() : null;
    },
    // Find meta description
    getMeta(name) {
      const match = html.match(new RegExp(`<meta[^>]+name=["']${name}["'][^>]+content=["']([^"']+)["']`, 'i')) ||
                   html.match(new RegExp(`<meta[^>]+content=["']([^"']+)["'][^>]+name=["']${name}["']`, 'i'));
      return match ? match[1] : null;
    },
    // Find all CSS links
    getCSSLinks() {
      const links = [];
      const regex = /<link[^>]+rel=["']stylesheet["'][^>]+href=["']([^"']+)["']/gi;
      let match;
      while ((match = regex.exec(html)) !== null) {
        links.push(match[1]);
      }
      // Also match href before rel
      const regex2 = /<link[^>]+href=["']([^"']+\.css[^"']*)["'][^>]*>/gi;
      while ((match = regex2.exec(html)) !== null) {
        if (!links.includes(match[1])) links.push(match[1]);
      }
      return links;
    },
    // Find all links matching pattern
    getLinks(pattern) {
      const links = [];
      const regex = new RegExp(`<a[^>]+href=["']([^"']*${pattern}[^"']*)["'][^>]*>`, 'gi');
      let match;
      while ((match = regex.exec(html)) !== null) {
        links.push(match[1]);
      }
      return [...new Set(links)];
    },
    // Find elements by class pattern
    getElementsByClass(pattern) {
      const regex = new RegExp(`<([a-z]+)[^>]+class=["'][^"']*${pattern}[^"']*["'][^>]*>`, 'gi');
      const elements = [];
      let match;
      while ((match = regex.exec(html)) !== null) {
        elements.push(match[1]);
      }
      return elements;
    },
    // Extract hex colors
    getColors() {
      const colors = new Set();
      const hexRegex = /#([0-9a-fA-F]{3,8})/g;
      let match;
      while ((match = hexRegex.exec(html)) !== null) {
        colors.add(match[0]);
      }
      // RGB colors
      const rgbRegex = /rgb\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)/g;
      while ((match = rgbRegex.exec(html)) !== null) {
        colors.add(`rgb(${match[1]}, ${match[2]}, ${match[3]})`);
      }
      return [...colors].slice(0, 30);
    },
    // Find main structural elements
    getStructure() {
      const structure = {
        header: this.getElementsByClass('header').length > 0,
        nav: html.includes('<nav'),
        main: html.includes('<main'),
        footer: html.includes('<footer'),
        sidebar: this.getElementsByClass('sidebar').length > 0,
      };
      return structure;
    },
  };
}

async function scrapeEporner() {
  const baseUrl = 'https://www.eporner.com';
  
  console.log('='.repeat(60));
  console.log('SCRAPING EPORNER.COM');
  console.log('='.repeat(60));

  try {
    // Scrape homepage
    console.log('\n📌 Fetching homepage...');
    const homepageHtml = await fetch(baseUrl);
    const homepage = parseHTML(homepageHtml);
    
    console.log(`\n✅ Page Title: ${homepage.getTitle()}`);
    console.log(`📝 Meta Description: ${homepage.getMeta('description')?.substring(0, 100)}...`);
    
    // CSS Files
    console.log('\n📄 CSS Files:');
    const cssFiles = homepage.getCSSLinks();
    cssFiles.slice(0, 10).forEach((css, i) => {
      console.log(`  ${i + 1}. ${css}`);
    });
    
    // Main navigation
    console.log('\n🔗 Main Navigation Links:');
    const navLinks = homepage.getLinks('/(home|videos|categories|pornstars|trending|popular)');
    navLinks.slice(0, 10).forEach(link => {
      console.log(`  - ${link}`);
    });
    
    // Video links
    console.log('\n🎬 Sample Video Links:');
    const videoLinks = homepage.getLinks('/video/');
    videoLinks.slice(0, 5).forEach(link => {
      console.log(`  - ${link}`);
    });
    
    // Category links
    console.log('\n📂 Category Links:');
    const categoryLinks = homepage.getLinks('/(cat|category)');
    categoryLinks.slice(0, 8).forEach(link => {
      console.log(`  - ${link}`);
    });
    
    // Colors found
    console.log('\n🎨 Colors found:');
    const colors = homepage.getColors();
    colors.slice(0, 15).forEach(color => {
      console.log(`  - ${color}`);
    });
    
    // Structure
    console.log('\n🏗️ Page Structure:');
    const structure = homepage.getStructure();
    Object.entries(structure).forEach(([key, value]) => {
      console.log(`  ${key}: ${value ? '✓' : '✗'}`);
    });
    
    // Try to find video page structure
    if (videoLinks.length > 0) {
      console.log('\n' + '='.repeat(60));
      console.log('SCRAPING VIDEO PAGE');
      console.log('='.repeat(60));
      
      const videoUrl = new URL(videoLinks[0], baseUrl).href;
      console.log(`\n📌 Fetching: ${videoUrl}`);
      
      try {
        const videoHtml = await fetch(videoUrl);
        const videoPage = parseHTML(videoHtml);
        
        console.log(`\n✅ Video Page Title: ${videoPage.getTitle()}`);
        
        // Video embed
        const iframeMatch = videoHtml.match(/<iframe[^>]+src=["']([^"']+)["']/i);
        if (iframeMatch) {
          console.log(`\n🎬 Video Embed: ${iframeMatch[1].substring(0, 80)}...`);
        }
        
        // Video info
        const titleMatch = videoHtml.match(/<h1[^>]*>([^<]+)<\/h1>/i);
        if (titleMatch) {
          console.log(`\n📌 Video Title: ${titleMatch[1].trim()}`);
        }
        
        // Views, rating patterns
        const viewsMatch = videoHtml.match(/(\d+(?:\.\d+)?[KM]?)\s*views?/i);
        if (viewsMatch) {
          console.log(`\n👁️ Views pattern: ${viewsMatch[1]}`);
        }
        
        const ratingMatch = videoHtml.match(/(\d+)%\s*(?:rating|rate)/i);
        if (ratingMatch) {
          console.log(`\n⭐ Rating pattern: ${ratingMatch[1]}%`);
        }
        
      } catch (e) {
        console.log(`\n⚠️ Could not fetch video page: ${e.message}`);
      }
    }
    
    // Generate CSS reference
    console.log('\n' + '='.repeat(60));
    console.log('GENERATING CSS REFERENCE');
    console.log('='.repeat(60));
    
    const reference = {
      baseUrl,
      cssFiles: cssFiles.slice(0, 15),
      colors: colors.slice(0, 20),
      structure,
      sampleVideoLinks: videoLinks.slice(0, 10),
      categoryLinks: categoryLinks.slice(0, 15),
    };
    
    // Save to file
    const fs = require('fs');
    fs.writeFileSync('eporner_reference.json', JSON.stringify(reference, null, 2));
    console.log('\n💾 Reference saved to: eporner_reference.json');
    
    console.log('\n' + '='.repeat(60));
    console.log('✅ SCRAPING COMPLETE');
    console.log('='.repeat(60));
    
  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

// Run the scraper
scrapeEporner();
